src/lgmres.m

```diff
@@ -0,0 +1,325 @@
+    if colsxInitial ~= 1
+        error("Initial guess xInitial is not a column vector.");
+    end
+
+    % Check whether x0 has the right dimension
+    if rowsxInitial ~= n
+        msg = "Dimension mismatch between matrix A and initial guess xInitial.";
+        error(msg);
+    end
+
+    clear rowsxInitial colsxInitial;
+
+    % ---> LGMRES Algorithm starts here
+    % First outer iteration is a simple restarted GMRES(m) execution
+    % First approximation error vector is created in this stage
+
+    % Algorithm setup
+    restart = 1;
+    r0 = b - A * xInitial;
+    res(1, :) = norm(r0);
+    relresvec(1, :) = (norm(r0) / res(1, 1));
+    iter(1, :) = restart;
+
+    % Matrix with the history of approximation error vectors
+    zMat = zeros(n, k);
+
+    % while number_of_cycles <=k, according to 'mi_lgmres',
+    % We execute GMRES(m+k) only
+
+    tic(); % start measuring CPU time
+
+    % Call MATLAB built-in gmres.
+    % Ref. [1], pag. 968, recommends GMRES(m+k)
+    % if no enough approximation error vectors are stored yet.
+    [x, flag, ~, ~, resvec] = ...
+        gmres(A, b, m + k, tol, 1, [], [], xInitial);
+
+    % Update residual norm, iterations, and relative residual vector
+    res(restart + 1, :) = resvec(end);
+    iter(restart + 1, :) = restart + 1;
+    relresvec(size(relresvec, 1) + 1, :) = resvec(end) / res(1, 1);
+    % First approximation error vector
+    zMat(:, restart) = x - xInitial;
+
+    % gmres uses a flag system. We only care wheter the solution has
+    % converged or not
+    if flag == 1 % if flag != 0?
+        flag = 0;
+        xInitial = x;
+        restart = restart + 1;
+    else
+        flag = 1;
+        time = toc();
+        return
+    end
+
+    % ---> LGMRES Algorithm for restart > 1 or
+    % LGMRES(m, k)
+
+    while flag == 0 && restart <= maxit
+
+        % Compute normalized residual vector
+        r = b - A * xInitial;
+        beta = norm(r);
+        v1 = r / beta;
+
+        % Modified Gram-Schmidt Arnoldi iteration
+        % Notice that for augmented Krylov subspaces, we need zHistory
+        % where zHistory, a n-by-k matrix, is the history of approximation
+        % error vectors from the last outer iterations
+        % For LGMRES we must take in consideration that
+        % s = m + k is related to the size of
+        % output parameteres H, V
+        [H, V, s] = ...
+            augmented_gram_schmidt_arnoldi ...
+            (A, v1, m + k - min(restart - 1, k), ...
+             zMat(:, 1:min(restart - 1, k)));
+
+        % Plane rotations
+        [HUpTri, g] = plane_rotations(H, beta);
+
+        % Solve the least-squares problem
+        % s instead of m
+        Rs = HUpTri(1:s, 1:s);
+        gs = g(1:s);
+        minimizer = Rs \ gs;
+        % For LGMRES, we should now consider the approximation error
+        % separately as a variable: zName can be zCurrent,
+        % zCycle, zCurrentCycle, name is open to suggestions
+
+        W = zeros(n, s);
```

Comment:
Please try to document what is going on here, i.e., from L282 to L288

Reply:
From Baker, fig. 1, lines 10 and 12, the n-by-s matrix W is created with the first m Arnoldi vectors and the k approximation error vectos, from newest to oldest. zCurrentCycle is the approximation error vector from the current outer iteration.